File: Problems/20.cpp
class Solution {
public:
    bool isValid(string s) {
        std::stack<char> st;

        for(auto elem: s){
            switch(elem){
                case '(':
                    st.push('('); break;
                case '{':
                    st.push('{'); break;
                case '[':
                    st.push('['); break;
                case ')':
                    if(st.empty()) return false;
                    if(st.top() != '(') return false;
                    st.pop(); break;
                case '}':
                    if(st.empty()) return false;
                    if(st.top() != '{') return false;
                    st.pop(); break;
                case ']':
                    if(st.empty()) return false;
                    if(st.top() != '[') return false;
                    st.pop(); break;
            }
        }

        return st.empty();

    }
};

File: Problems/20.rs
impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut stack: Vec<char> = vec![];

        for e in s.chars() {
            match e {
                '(' => {
                    stack.push('(');
                }
                ')' => {
                    let top = stack.pop();

                    if let Some(x) = top {
                        if x != '(' {
                            return false;
                        }
                    } else {
                        return false;
                    }
                }
                '[' => {
                    stack.push('[');
                }
                ']' => {
                    let top = stack.pop();

                    if let Some(x) = top {
                        if x != '[' {
                            return false;
                        }
                    } else {
                        return false;
                    }
                }
                '{' => {
                    stack.push('{');
                }
                _ => {
                    let top = stack.pop();

                    if let Some(x) = top {
                        if x != '{' {
                            return false;
                        }
                    } else {
                        return false;
                    }
                }
            }
        }

        if stack.len() == 0 {
            return true;
        }

        false
    }
}
